File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/ServerAsync/QueryVatTuTheoMauAsycn.java
package vinhlong.ditagis.com.khaosatdongho.async.ServerAsync;

import android.app.Activity;
import android.os.AsyncTask;
import android.support.design.widget.BottomSheetDialog;
import android.util.Log;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

import vinhlong.ditagis.com.khaosatdongho.R;
import vinhlong.ditagis.com.khaosatdongho.utities.Constant;


public class QueryVatTuTheoMauAsycn extends AsyncTask<String, ArrayList<QueryVatTuTheoMauAsycn.VatTu>, Void> {
    private BottomSheetDialog mDialog;
    private Activity mActivity;
    private AsyncResponse mDelegate;

    public interface AsyncResponse {
        void processFinish(ArrayList<VatTu> output);
    }

    public QueryVatTuTheoMauAsycn(Activity activity, AsyncResponse delegate) {
        this.mActivity = activity;
        this.mDelegate = delegate;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        mDialog = new BottomSheetDialog(this.mActivity);
        LinearLayout view = (LinearLayout) mActivity.getLayoutInflater().inflate(R.layout.layout_progress_dialog, null, false);
        ((TextView) view.findViewById(R.id.txt_progress_dialog_title)).setText("Đang lấy danh sách tên mẫu...");
        mDialog.setContentView(view);
        mDialog.setCancelable(false);

        mDialog.show();
    }

    @Override
    protected Void doInBackground(String... params) {
        try {
            if (params != null && params.length > 0) {
                String tenThietLapMau = params[0];
                String urlAPI = String.format(Constant.API_URL.VATTU_THEOMAU, tenThietLapMau);
                URL url = new URL(urlAPI);
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                try {
                    conn.setDoOutput(false);
                    conn.setRequestMethod(Constant.METHOD.GET);
                    conn.connect();
                    BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                    StringBuffer buffer = new StringBuffer();
                    String line;
                    while ((line = bufferedReader.readLine()) != null) {
                        buffer.append(line);
                    }
                    ArrayList<VatTu> vatTus = getVatTus(buffer.toString());
                    publishProgress(vatTus);
                    this.mDialog.dismiss();
                } catch (Exception e) {
                    Log.e("error", e.toString());
                } finally {
                    conn.disconnect();
                }
            }
        } catch (Exception e) {
            Log.e("ERROR", e.toString());
        }
        return null;
    }
    private ArrayList<VatTu> getVatTus(String data) throws JSONException {
        ArrayList<VatTu> vatTus = new ArrayList<>();
        if (data != null) {
            JSONObject jsonData = new JSONObject(data);
            JSONArray jsonArray = jsonData.getJSONArray("value");
            for (int i = 0; i < jsonArray.length(); i++) {
                VatTu vatTu = new VatTu();
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                Object maVatTu = jsonObject.get("MaVatTu");
                if (maVatTu != null) {
                    vatTu.setMaVatTu(maVatTu.toString());
                }
                Object tenMau = jsonObject.get("TENMAU");
                if (tenMau != null) {
                    vatTu.setTenMau(tenMau.toString());
                }
                Object soLuong = jsonObject.get("SoLuong");
                if (soLuong != null) {
                    vatTu.setSoLuong(soLuong.toString());
                }
                Object id = jsonObject.get("ID");
                if (id != null) {
                    vatTu.setId(id.toString());
                }
                vatTus.add(vatTu);
            }
        }
        return vatTus;
    }
    @Override
    protected void onProgressUpdate(ArrayList<VatTu>... values) {
        super.onProgressUpdate(values);
        if (values == null || values.length == 0)
            this.mDelegate.processFinish(null);
        this.mDelegate.processFinish(values[0]);

    }

    @Override
    protected void onPostExecute(Void value) {
    }
    public class VatTu{
        private String maVatTu;
        private String tenMau;
        private String soLuong;
        private String id;

        public VatTu() {
        }

        public VatTu(String maVatTu, String tenMau, String soLuong, String id) {
            this.maVatTu = maVatTu;
            this.tenMau = tenMau;
            this.soLuong = soLuong;
            this.id = id;
        }

        public String getMaVatTu() {
            return maVatTu;
        }

        public void setMaVatTu(String maVatTu) {
            this.maVatTu = maVatTu;
        }

        public String getTenMau() {
            return tenMau;
        }

        public void setTenMau(String tenMau) {
            this.tenMau = tenMau;
        }

        public String getSoLuong() {
            return soLuong;
        }

        public void setSoLuong(String soLuong) {
            this.soLuong = soLuong;
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }
    }
}

File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/ServerAsync/QueryVatTuTheoMauAsycn.kt
package vinhlong.ditagis.com.khaosatdongho.async.ServerAsync

import android.app.Activity
import android.os.AsyncTask
import android.util.Log
import org.json.JSONException
import org.json.JSONObject
import vinhlong.ditagis.com.khaosatdongho.utities.Constant
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.*


class QueryVatTuTheoMauAsycn(private val mActivity: Activity, private val mDelegate: AsyncResponse) : AsyncTask<String, ArrayList<QueryVatTuTheoMauAsycn.VatTu>, Void>() {

    interface AsyncResponse {
        fun processFinish(vatTus: ArrayList<VatTu>?)
    }

    override fun doInBackground(vararg params: String): Void? {
        try {
            if (params != null && params.isNotEmpty()) {
                val tenThietLapMau = params[0]
                val urlAPI = String.format(Constant.API_URL.VATTU_THEOMAU, tenThietLapMau)
                val url = URL(urlAPI)
                val conn = url.openConnection() as HttpURLConnection
                try {
                    conn.doOutput = false
                    conn.requestMethod = Constant.METHOD.GET
                    conn.connect()
                    val bufferedReader = BufferedReader(InputStreamReader(conn.inputStream))
                    val buffer = StringBuffer()
                    var line: String?
                    while (true) {
                        line = bufferedReader.readLine()
                        if(line == null) break
                        buffer.append(line)
                    }
                    val vatTus = getVatTus(buffer.toString())
                    publishProgress(vatTus)
                } catch (e: Exception) {
                    Log.e("error", e.toString())
                } finally {
                    conn.disconnect()
                }
            }
        } catch (e: Exception) {
            Log.e("ERROR", e.toString())
        }

        return null
    }

    @Throws(JSONException::class)
    private fun getVatTus(data: String?): ArrayList<VatTu> {
        val vatTus = ArrayList<VatTu>()
        if (data != null) {
            val jsonData = JSONObject(data)
            val jsonArray = jsonData.getJSONArray("value")
            for (i in 0 until jsonArray.length()) {
                val vatTu = VatTu()
                val jsonObject = jsonArray.getJSONObject(i)
                val maVatTu = jsonObject.get("MaVatTu")
                if (maVatTu != null) {
                    vatTu.maVatTu = maVatTu.toString()
                }
                val tenMau = jsonObject.get("TENMAU")
                if (tenMau != null) {
                    vatTu.tenMau = tenMau.toString()
                }
                val soLuong = jsonObject.get("SoLuong")
                if (soLuong != null) {
                    vatTu.soLuong = soLuong as Double
                }
                val id = jsonObject.get("ID")
                if (id != null) {
                    vatTu.id = id.toString()
                }
                vatTus.add(vatTu)
            }
        }
        return vatTus
    }

    override fun onProgressUpdate(vararg values: ArrayList<VatTu>) {
        super.onProgressUpdate(*values)
        if (values == null || values.size == 0)
            this.mDelegate.processFinish(null)
        this.mDelegate.processFinish(values[0])

    }

    inner class VatTu {
        var maVatTu: String? = null
        var tenMau: String? = null
        var soLuong: Double = 0.toDouble()
        var id: String? = null

        constructor()
    }
}
